gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/api/use_case/MigrateApiUseCase.java

```diff
@@ -75,39 +81,57 @@ public MigrateApiUseCase(
             apiIndexerDomainService,
             apiPrimaryOwnerDomainService,
             planService,
+            flowCrudService,
             apiStateService,
             new V2toV4MigrationOperator()
         );
     }
 
     public Output execute(Input input) {
-        var api = apiCrudService.findById(input.apiId());
-        // Preconditions
-        if (api.isEmpty()) {
-            throw new ApiNotFoundException(input.apiId());
-        }
+        var api = apiCrudService.findById(input.apiId()).orElseThrow(() -> new ApiNotFoundException(input.apiId()));
         MigrationResult<?> precondition = MigrationResult.value(1);
-        if (api.get().getDefinitionVersion() != DefinitionVersion.V2) {
+        if (api.getDefinitionVersion() != DefinitionVersion.V2) {
+            // Fatal issue, we don’t try to do anything in this case
             return new Output(
                 input.apiId(),
                 List.of(new MigrationResult.Issue("Cannot upgrade an API which is not a v2 definition", MigrationResult.State.IMPOSSIBLE))
             );
-        } else if (!apiStateService.isSynchronized(api.get(), input.auditInfo())) {
+        }
+        if (!apiStateService.isSynchronized(api, input.auditInfo())) {
             precondition =
                 precondition.addIssue(
                     new MigrationResult.Issue("Cannot upgrade an API which is out of sync", MigrationResult.State.CAN_BE_FORCED)
                 );
         }
+        if (api.getApiDefinition().getExecutionMode() == ExecutionMode.V3) {
+            precondition =
+                precondition.addIssue(
+                    new MigrationResult.Issue("Cannot upgrade an API which isn’t in V4 emulation", MigrationResult.State.IMPOSSIBLE)
```

Comment:
Instead of `upgrade` I would use `migrate` in the messages.

```suggestion
                    new MigrationResult.Issue("Cannot migrate an API not using V4 emulation", MigrationResult.State.IMPOSSIBLE)
```